crud/borders.lua

```diff
@@ -0,0 +1,177 @@
+local checks = require('checks')
+local errors = require('errors')
+local vshard = require('vshard')
+
+local dev_checks = require('crud.common.dev_checks')
+local call = require('crud.common.call')
+local utils = require('crud.common.utils')
+local schema = require('crud.common.schema')
+local Keydef = require('crud.compare.keydef')
+
+local BorderError = errors.new_class('Border',  {capture_stack = false})
+
+local borders = {}
+
+local STAT_FUNC_NAME = '_crud.get_border_on_storage'
+
+
+local function get_border_on_storage(border_name, space_name, index_name, field_names)
+    dev_checks('string', 'string', 'string', '?table')
+
+    assert(border_name == 'min' or border_name == 'max')
+
+    local space = box.space[space_name]
+    if space == nil then
+        return nil, BorderError:new("Space %q doesn't exist", space_name)
+    end
+
+    local index = space.index[index_name]
+    if index == nil then
+        return nil, BorderError:new("Index %q of space doesn't exist", index_name, space_name)
+    end
+
+    local function get_border(index)
+        return index[border_name](index)
+    end
+
+    return schema.wrap_func_result(space, get_border, {index}, {
+        add_space_schema_hash = true,
+        field_names = field_names,
+    })
+end
+
+function borders.init()
+   _G._crud.get_border_on_storage = get_border_on_storage
+end
+
+local function is_closer(border_name, keydef, tuple, res_tuple)
+    assert(border_name == 'min' or border_name == 'max')
+
+    if res_tuple == nil then
+        return true
+    end
+
+    local cmp = keydef:compare(tuple, res_tuple)
+
+    return border_name == 'min' and cmp < 0 or border_name == 'max' and cmp > 0
+end
+
+local function get_border(border_name, space_name, index_name, opts)
+    checks('string', 'string', '?string', {
```

Comment:
`index_name` could be also a number.

Usually I use `space.index[0]:min()`. I don't use index name but I know that 0 is a primary key.